datafusion/core/src/physical_plan/coalesce_partitions.rs

```diff
@@ -199,6 +200,28 @@ impl Stream for MergeStream {
         cx: &mut std::task::Context<'_>,
     ) -> Poll<Option<Self::Item>> {
         let poll = self.input.poll_recv(cx);
+
+        // If the input stream is done, wait for all tasks to finish and return
+        // the failure if any.
+        if let Poll::Ready(None) = poll {
+            let fut = self.tasks.join_next();
```

Comment:
It seems like this only checks the first future that finishes. Would it work if there were multiple input tasks and the first one that finished was not the one that `panic`d?

Reply:
BTW I have been working on writing a test for this as part of https://github.com/apache/arrow-datafusion/pull/6507